lasagne/tests/layers/test_recurrent.py

```diff
@@ -94,15 +94,78 @@ def test_recurrent_init_val_error():
         l_rec = RecurrentLayer(InputLayer((2, 2, 3)), 5, hid_init=hid_init)
 
 
+def test_custom_recurrent_arbitrary_shape():
+    # Check that the custom recurrent layer can handle more than 1 feature dim
+    n_batch, n_steps, n_channels, width, height = (2, 3, 4, 5, 6)
+    n_out_filters = 7
+    filter_shape = (3, 3)
+    l_in = lasagne.layers.InputLayer(
+        (n_batch, n_steps, n_channels, width, height))
+    l_in_to_hid = lasagne.layers.Conv2DLayer(
+        lasagne.layers.InputLayer((None, n_channels, width, height)),
+        n_out_filters, filter_shape, pad='same')
+    l_hid_to_hid = lasagne.layers.Conv2DLayer(
+        lasagne.layers.InputLayer((None, n_out_filters, width, height)),
+        n_out_filters, filter_shape, pad='same')
+    l_rec = lasagne.layers.CustomRecurrentLayer(
+        l_in, l_in_to_hid, l_hid_to_hid)
+    out = theano.function([l_in.input_var], lasagne.layers.get_output(l_rec))
+    out_shape = out(np.zeros((n_batch, n_steps, n_channels, width, height),
+                             dtype=theano.config.floatX)).shape
+    assert out_shape == (n_batch, n_steps, n_out_filters, width, height)
```

Comment:
In addition, you should check that `l_rec.output_shape` matches the expected shape. And you will see that it doesn't :) (I had overlooked this until now! `get_output_shape_for()` still accesses `self.num_units` instead of `self.hid_shape`. And maybe `self.hid_shape` isn't that useful anyway, maybe `get_output_shape_for()` should just use `self.l_hid_to_hid.output_shape[1:]` directly, and the constructor should use `self.output_shape` or `self.l_hid_to_hid.output_shape`, depending on what's needed.)